monai/transforms/spatial/array.py

```diff
@@ -310,21 +310,27 @@ def __call__(
             spatial_ornt = src
         else:
             if self.axcodes is None:
-                raise AssertionError
+                raise ValueError("Incompatible values: axcodes=None and as_closest_canonical=True.")
             dst = nib.orientations.axcodes2ornt(self.axcodes[:sr], labels=self.labels)
             if len(dst) < sr:
                 raise ValueError(
                     f"axcodes must match data_array spatially, got axcodes={len(self.axcodes)}D data_array={sr}D"
                 )
             spatial_ornt = nib.orientations.ornt_transform(src, dst)
-        ornt = spatial_ornt.copy()
-        ornt[:, 0] += 1  # skip channel dim
-        ornt = np.concatenate([np.array([[0, 1]]), ornt])
-        shape = data_array_np.shape[1:]
-        data_array_np = np.ascontiguousarray(nib.orientations.apply_orientation(data_array_np, ornt))
-        new_affine = affine_ @ nib.orientations.inv_ornt_aff(spatial_ornt, shape)
+        new_affine = affine_ @ nib.orientations.inv_ornt_aff(spatial_ornt, spatial_shape)
+        _is_tensor = isinstance(data_array, torch.Tensor)
+        spatial_ornt[:, 0] += 1  # skip channel dim
+        spatial_ornt = np.concatenate([np.array([[0, 1]]), spatial_ornt])
+        axes = [ax for ax, flip in enumerate(spatial_ornt[:, 1]) if flip == -1]
+        if axes:
+            data_array = torch.flip(data_array, dims=axes) if _is_tensor else np.flip(data_array, axis=axes)  # type: ignore
+        full_transpose = np.arange(len(data_array.shape))
+        full_transpose[: len(spatial_ornt)] = np.argsort(spatial_ornt[:, 0])
+        data_array = (
+            data_array.permute(full_transpose.tolist()) if _is_tensor else data_array.transpose(full_transpose)  # type: ignore
+        )
         new_affine = to_affine_nd(affine_np, new_affine)
-        out, *_ = convert_to_dst_type(src=data_array_np, dst=data_array)
+        out = data_array.contiguous() if _is_tensor else np.ascontiguousarray(data_array)  # type: ignore
```

Comment:
I think you can simplify it to:
```py
out, *_ = convert_to_dst_type(src=data_array, dst=data_array)
```
The `ToTensor` and `ToNumpy` logic already have the `contiguous` conversion:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/type_conversion.py#L173
https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/type_conversion.py#L120

Thanks.

Reply:
thanks, I updated the PR and added logic to skip the transpose if not needed